contracts/adapters/GmpTransferAdapter.sol

```diff
@@ -0,0 +1,147 @@
+            revert InsufficientMsgValueAmount(msg.value);
+        // refund excess msg.value
+        } else if (msg.value > fee) {
+            (bool success, ) = msg.sender.call{value: msg.value - fee}("");
+            if (!success) revert FailedRefund();
+        }
+
+        address destinationToken;
+        address assignedDestinationToken = crossChainTokenPairs[XERC20Address][destinationDomainID];
+        if (assignedDestinationToken != address(0)) {
+            destinationToken = assignedDestinationToken;
+        } else {
+            destinationToken = XERC20Address;
+        }
+
+        bytes memory depositData = abi.encodePacked(
+            // uint256 maxFee
+            uint256(950000),
+            // uint16 len(executeFuncSignature)
+            uint16(4),
+            // bytes executeFuncSignature
+            IGmpTransferAdapter(address(this)).executeProposal.selector,
+            // uint8 len(executeContractAddress)
+            uint8(20),
+            // bytes executeContractAddress
+            address(this),
+            // uint8 len(executionDataDepositor)
+            uint8(20),
+            // bytes executionDataDepositor
+            address(this),
+            // bytes executionDataDepositor + executionData
+            prepareDepositData(recipientAddress, destinationToken, tokenAmount)
+        );
+
+        IXERC20(XERC20Address).burn(msg.sender, tokenAmount);
+
+        _bridge.deposit{value: fee}(destinationDomainID, _resourceID, depositData, "");
+    }
+
+    function executeProposal(address gmpAdapter, address recipient, address XERC20Address, uint256 amount) external {
+        if (gmpAdapter != address(this)) revert InvalidOriginAdapter(gmpAdapter);
+        if (msg.sender != _gmpAddress) revert InvalidHandler(msg.sender);
+
+        IXERC20(XERC20Address).mint(recipient, amount);
+    }
+
+    /**
+        @notice Used to manually transfer native tokens from Adapter.
+        @param recipient Address that should recieve the native tokens.
+        recipient   address
+     */
+    function withdraw(address recipient, uint256 amount) external onlyAdmin {
+        (bool success, ) = address(recipient).call{value: amount}("");
+        if(!success) revert FailedFundsTransfer();
+        emit Withdrawal(recipient, amount);
+    }
+
+    function setTokenPairAddress(address sourceTokenAddress, uint8 destinationDomainID, address destinationTokenAddress) external onlyAdmin {
+        crossChainTokenPairs[sourceTokenAddress][destinationDomainID] = destinationTokenAddress;
+    }
+
+    function slice(bytes calldata input, uint256 position) public pure returns (bytes memory) {
+        return input[position:];
+    }
+
+    function prepareDepositData(
+        address recipientAddress,
+        address XERC20Address,
+        uint256 bridgingAmount
+    ) public view returns (bytes memory) {
+        return abi.encode(recipientAddress, XERC20Address, bridgingAmount);
+    }
+
+    receive() external payable {}
```

Comment:
I think there is no use case where someone is sending native currency to the adapter directly.
```suggestion
```